mlflow/llama_index/serialize_objects.py

```diff
@@ -0,0 +1,183 @@
+import importlib
+import inspect
+import json
+import logging
+from typing import Callable, Dict
+
+import llama_index
+from llama_index.core import PromptTemplate
+
+_logger = logging.getLogger(__name__)
+
+OBJECT_DICT_METHOD_MAP = {
+    llama_index.core.llms.llm.LLM: ("to_dict", "from_dict"),
+    llama_index.core.base.embeddings.base.BaseEmbedding: ("to_dict", "from_dict"),
+    llama_index.core.node_parser.interface.NodeParser: ("to_dict", "from_dict"),
+    llama_index.core.indices.prompt_helper.PromptHelper: ("to_dict", "from_dict"),
+}
+
+
+def _object_to_dict(obj: object) -> Dict[str, any]:
+    for k, v in OBJECT_DICT_METHOD_MAP.items():
+        if isinstance(obj, k):
+            if not hasattr(obj, v[0]):
+                raise AttributeError(
+                    f"Object {obj} was inferred to be of type {k} but does not "
+                    "have a {v[0]} method. Ensure that `OBJECT_DICT_METHOD_MAP` is "
+                    "correct and the object is of type {k}."
+                )
+
+            return getattr(obj, v[0])()
+
+    raise AttributeError(f"Object {obj} does not have a supported deserialization method.")
+
+
+def _get_object_import_path(o: object, do_validate_import: bool = False) -> str:
+    """Return an import path to the object."""
+    if not inspect.isclass(o):
+        o = o.__class__
+
+    module_name = inspect.getmodule(o).__name__
+    class_name = o.__qualname__
+
+    if do_validate_import:
+        module = importlib.import_module(module_name)
+
+        if not hasattr(module, class_name):
+            raise ValueError(f"Module {module} does not have {class_name}")
+
+    return f"{module_name}.{class_name}"
+
+
+def _sanitize_api_key(object_as_dict: Dict[str, str]) -> Dict[str, str]:
+    keys_to_remove = [k for k in object_as_dict.keys() if "api_key" in k.lower()]
+
+    for k in keys_to_remove:
+        if object_as_dict.pop(k, None):
+            _logger.info(
+                "API key removed from object serialization. At inference time,"
+                " the key must be passed as an environment variable or via inference"
+                " parameters."
```

Comment:
nit: I'd remove `or via inference parameters`.... it will highly likely create a security hole e.g. put API key in the code😅